<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="canonical" href="https://xuanblo.github.io/NGSPipeDb/NGSPipe-RNA-seq/">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>NGSPipe RNA-seq - NGSPipeDb Document</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "NGSPipe RNA-seq";
    var mkdocs_page_input_path = "NGSPipe-RNA-seq.md";
    var mkdocs_page_url = "/NGSPipeDb/NGSPipe-RNA-seq/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> NGSPipeDb Document</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Getting start</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSPipeDb/">NGSPipeDb Tutorial</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">NGSPipe RNA-seq</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#quick-start-one-time-installation-of-components-necessary-for-an-individual-user">Quick Start - One time installation of components necessary for an individual user </a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#step-by-step-rna-seq-workflow">Step-by-step RNA-seq workflow </a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#1-installing-wget-and-git">1. Installing wget and git </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#2-installing-miniconda3">2. Installing Miniconda3 </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#3-downloading-the-ngspipedb-project">3. Downloading the NGSPipeDb project </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#4-installing-the-ngspipedb-conda-environments">4. Installing the NGSPipeDb conda environments </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#5-downloading-the-test-files">5. Downloading the test files </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#6-run-rna-seq-analysis-on-test-data">6. run RNA-seq analysis on test data </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#7-rgenerate-report">7. rgenerate report </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#8-run-your-custome-data">8. run your custome data </a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#rawdata-sequence-data">Rawdata sequence data</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#reference-data">reference data</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#edit-config-file">edit config file</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#condition-for-sample-compare">condition for sample compare</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#edit-snakefile">edit snakefile</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#report">report</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#run-snakemake">run snakemake</a>
    </li>
        </ul>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSPipe-ChIP-seq/">NGSPipe ChIP-seq</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSPipe-resequecing/">NGSPipe resequecing</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSDb/">NGSDb</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../linux/">Linux basic</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">NGSPipeDb Document</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>NGSPipe RNA-seq</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/xuanblo/NGSPipeDb/mkdocs/edit/master/docs/NGSPipe-RNA-seq.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="rna-seq-analysis">RNA-seq analysis</h1>
<p><strong>Table of Contents:</strong></p>
<ol>
<li><a href="#BasicLinux">Installing wget and git</a></li>
<li><a href="#Miniconda">Installing Miniconda3</a></li>
<li><a href="#NGSPipeDbEnv">Installing the NGSPipeDb conda environments</a></li>
<li><a href="#NGSPipeDbSource">Downloading the NGSPipeDb source code</a></li>
<li><a href="#Testdata">Downloading the NGSPipeDb test files</a></li>
<li><a href="#RunTest">Run test data</a></li>
<li><a href="#Report">Generate report</a></li>
<li><a href="#RunRawdata">Run your custome data</a></li>
</ol>
<h2 id="quick-start-one-time-installation-of-components-necessary-for-an-individual-user">Quick Start - One time installation of components necessary for an individual user <a name="QuickStarted"></a></h2>
<p>Three commands to start analysing <a href="">test data</a>:</p>
<pre><code class="language-shell"># download ngspipedb to anywhere you want
git clone git://github.com/xuanblo/NGSPipeDb.git &amp;&amp; mv NGSPipeDb mouse_transcriptome_analysis &amp;&amp; cd mouse_transcriptome_analysis
# download test data and create environment
bash ngspipe/scripts/one_step_ranseq_test.sh
</code></pre>
<p>Now you can viste your website on http://127.0.0.1:8000. All result are stored in <code>results</code>.
- Example of report <sub><a href="http://www.liu-lab.com"><img alt="html" src="https://img.icons8.com/ios/20/000000/html-filetype.png" /></a></sub>.
- Example of database <sub><a href="http://www.liu-lab.com"><img alt="html" src="https://img.icons8.com/dotty/25/000000/copy-link.png" /></a></sub>.</p>
<p>If you have more time, then we recommend you configure ngspipedb according to your needs. For more details, please see <a href="#step-by-step">step by step</a> bellow.</p>
<h2 id="step-by-step-rna-seq-workflow">Step-by-step RNA-seq workflow <a href="Step-by-Step-RNASeq"></a></h2>
<p>Although included in this section are step-by-step instructions, it is assumed that the user has a basic understanding of the <a href="https://en.wikipedia.org/wiki/Command-line_interface">nix command line interface</a>. Also, best practice RNA-seq analysis is plus.</p>
<h3 id="1-installing-wget-and-git">1. Installing wget and git <a name="BasicLinux"></a></h3>
<p>To get some of the required software packages, we will use the command line tools called <a href="http://www.gnu.org/software/wget/">wget</a> and <a href="https://git-scm.com/book/en/v2/Getting-Started-Installing-Git">git</a>.  <em>wget</em> is a popular tool for downloading things off of the internet.  <em>git</em> is a distributed version control system which we will use to checkout the NGSPipeDb code.</p>
<p><strong>Note</strong>: These tools are already pre-installed in most systems, but if you are unsure whether or not you have <em>wget</em> enter <code>wget</code> and if the return is <code>wget: command not found</code>, then you will have to install <em>wget</em>.  Do likewise for <em>git</em>.</p>
<h3 id="2-installing-miniconda3">2. Installing Miniconda3 <a name="Miniconda"></a></h3>
<p>NGSPipeDb relies on the conda package manager for installation and dependency resolution, so you will need to install <a href="https://conda.io/docs/user-guide/install/index.html">conda</a> first.</p>
<p>We will be using the <a href="http://conda.pydata.org/miniconda.html">Miniconda3</a> package management system (aka <strong>CONDA</strong>) to manage all of the software packages that <strong>NGSPipeDb</strong> is dependent on. </p>
<p>Use following commands to retrieve and then run the Minicoda3 installation script:
1. <code>wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh</code><br />
2. <code>bash Miniconda3-latest-Linux-x86_64.sh</code><br />
    - While running the installation script, follow the commands listed on screen, and press the <em>enter</em> key to scroll. <strong>Make sure to answer <code>yes</code> when asked if you want to prepend Miniconda3 to PATH.</strong>
    - Close your terminal, open a new one and you should now have Conda working! You could, alternatively, run <code>source ~/.bashrc</code> to initiate conda.
    - Test by entering: <code>conda update conda</code>
        - Press <code>y</code> to confirm the conda updates
3. Mamba is a reimplementation of the conda package manager in C++, the fast conda-alternative. If you already have conda: 
    - <code>conda install mamba -c conda-forge</code></p>
<p><strong>Note</strong>: you will only have to install Minicoda3 once.</p>
<h3 id="3-downloading-the-ngspipedb-project">3. Downloading the NGSPipeDb project <a name="NGSPipeDbSource"></a></h3>
<p>To install the latest stable version of NGSPipeDb, please clone the git repository to your system.</p>
<pre><code>cd /path/to/where_you_want
git clone https://www.github.com/xuanblo/NGSPipeDb
</code></pre>
<p>or </p>
<pre><code>cd /path/to/where_you_want
git clone git://www.github.com/xuanblo/NGSPipeDb
</code></pre>
<p>If you want to use specific version, please cleckout the <a href="https://github.com/xuanblo/NGSPipeDb/releases">release</a> and issue the following commands:</p>
<pre><code>cd /path/to/where_you_want
wget https://github.com/xuanblo/NGSPipeDb/archive/NGSPipeDb_xxx.tar.gz
tar -xf NGSPipeDb_xxx.tar.gz
</code></pre>
<h3 id="4-installing-the-ngspipedb-conda-environments">4. Installing the NGSPipeDb conda environments <a name="NGSPipeDbEnv"></a></h3>
<p>We are now ready to use conda to install the software packages which NGSPipeDb is dependent on.</p>
<p>First, you will need to create the conda environments needed by the various workflows. The following command will create a new conda environment containing Snakemake and python called "ngspipedb" into which conda is installed, the default environment path is <code>~/miniconda/conda/env/ngspipedb</code>.</p>
<pre><code>mamba create -c conda-forge -c bioconda --name ngspipedb snakemake python=3.8
</code></pre>
<p>Next, to analysis NGS data some bioinformatics tools need to be installed.</p>
<pre><code>mamba env update -n ngspipedb --file ngspipe/envs/requirement.yaml --prune
</code></pre>
<p><strong>Note</strong>: By default, all dependencies and tools should automatically be installed on the first execution. In case of download problems, please see how to <a href="">install conda env local</a>, <a href="">share conda env</a> for more detail.</p>
<p><font color='red'>由于差异表达需要安装的包与前面的环境有不兼容，所以暂时独立了出来。</font></p>
<pre><code class="language-shell">conda create -n exp_r_env mamba -c conda-forge
conda activate exp_r_env
mamba env update --file ngspipe/envs/requirements_exp_r_env.yaml --prune
</code></pre>
<h3 id="5-downloading-the-test-files">5. Downloading the test files <a name="Testdata"></a></h3>
<p><strong>NGSPipeDb</strong> is dependent on reference files which can be found for the supported species listed below: <a href="http://www.liu-lab.com/ngspipedb">download link</a></p>
<p>To download RNA-seq test data into <code>./testdata</code>, simply run the bash script <code>download_testdata.sh</code>:</p>
<pre><code class="language-shell">cd NGSPipeDb
bash ngspipe/scripts/download_testdata.sh testdata
</code></pre>
<p>or you download the reference files that you need and then untarring then in a directory called <code>testdata</code>.</p>
<pre><code class="language-shell">wget http://www.liu-lab.com/ngspipedb/rnaseq_testdata.tar.gz
tar -zxvf rnaseq_testdata.tar.gz
</code></pre>
<p>Make sure you have the following directory structure by <code>tree testdata</code>:</p>
<pre><code>testdata/
├── GRCm38.83.chr19.gtf
├── chr19.fa
├── control_R1.fq.gz
├── control_R2.fq.gz
├── samples.xls
├── treated_R1.fq.gz
└── treated_R2.fq.gz
</code></pre>
<h3 id="6-run-rna-seq-analysis-on-test-data">6. run RNA-seq analysis on test data <a name="RunTest"></a></h3>
<p>We provied a simple RNA-seq workflow for you to take a glance of NGSPipe. In RNA-seq analysis part, it contains 7 step analysis:</p>
<pre><code class="language-python">1. sampling data
2. raw reads qc
3. junction alignmnet
4. transcript assembly
5. quantification
6. statistic
7. diff
</code></pre>
<p>First, we need to check the <code>ngspipe/file</code>. This will not execute anything, but display what would be done.</p>
<pre><code>snakemake -s ngspipe/workflow/rnaseq_analysis.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -np
</code></pre>
<p>If nothing goes wrong, you can generate a dag plot:</p>
<pre><code>snakemake -s ngspipe/workflow/rnaseq_analysis.Snakefile.py --dag|dot -Tpng &gt; dag.png
</code></pre>
<p>Now you can do RNA-seq analysis by just one simply command.</p>
<pre><code>snakemake -s ngspipe/workflow/rnaseq_analysis.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -p -j 10
</code></pre>
<p>The final data files are put in the folder <code>results</code>. Please check you result file <code>tree -d -L 2 results/</code>, it may like this:</p>
<pre><code>results/
├── report
│   ├── 1.rawreads_stat
│   ├── 2.cleanreads_stat
│   ├── 3.mapping_stat
│   ├── 4.exp_stat
│   └── workflow
├── result
│   ├── junction_align
│   ├── quantify
│   ├── rawReads_qc
│   ├── sampling_data
│   ├── statistic
│   └── transcript_assembly
└── sqlite3
    ├── blastdb
    ├── exp
    ├── gbrowse
    └── gff_sqlite3
</code></pre>
<p><strong>Note</strong>: If you encounter any problem in this step, please turn to <code>TroubleShooting</code> for help.</p>
<h3 id="7-rgenerate-report">7. rgenerate report <a name="Report"></a></h3>
<p>If all goes well, the proper analysis will be followed by the making of the html report using Snakemake to a html report file with pictures and tables.</p>
<pre><code>snakemake --snakefile ngspipe/workflow/rnaseq_analysis.Snakefile.py --report results/Report/report.html
</code></pre>
<p>The final report should appear as <code>results/Report/report.html</code>. This report is a single html file with all in it and can be sent to customers/colleagues as a final report. It is nicer than a PDF version because of large tables and figures which would suffer from page breaks and it can be viewed on any device supporting html include smartphones :-).</p>
<p><strong>Note</strong>: Internet Explorer is not supported. If you get connected error in this step, you can solve this problem by edit file <code>ngspipedb_py38_conda_env/lib/python3.8/site-packages/snakemake/report/report.html.jinja2</code> to change <code>https://raw.githubusercontent.com/eligrey/FileSaver.js/2.0.0/src/FileSaver.js</code> to <code>https://cdnjs.cloudflare.com/ajax/libs/FileSaver.js/2.0.0/FileSaver.js</code></p>
<h3 id="8-run-your-custome-data">8. run your custome data <a name="RunRawdata"></a></h3>
<p>NGSPipe is built to be used routinely. To ensure a maximum comparability of the results, you can copy default <code>ngspipe/config/rnaseq.config.yaml</code> and <code>ngspipe/workflow/rnaseq_analysis.Snakefile.py</code> file from the ngspipe source directory:</p>
<pre><code class="language-shell">cp ngspipe/config/rnaseq.config.yaml ngspipe/workflow/rnaseq_analysis.Snakefile.py ./
</code></pre>
<p>We will explain how to edit and configure these files shortly below.</p>
<h4 id="rawdata-sequence-data">Rawdata sequence data</h4>
<p>Raw data files can either be fastq, fastq.gz formated files. If your raw data are located in <strong>somewhere else</strong>, you can copy them to <code>rawdata</code>, or create soft links like <code>ln -s ../yoursamplepath/*.fq.gz rawdata/</code>.</p>
<pre><code>rawdata/
├── lung-rep1_R1.fq.gz
├── lung-rep1_R2.fq.gz
├── lung-rep2_R1.fq.gz
├── lung-rep2_R2.fq.gz
├── lung-rep3_R1.fq.gz
├── lung-rep3_R2.fq.gz
├── liver-rep1_R1.fq.gz
├── liver-rep1_R2.fq.gz
├── liver-rep2_R1.fq.gz
├── liver-rep2_R2.fq.gz
├── liver-rep3_R1.fq.gz
├── liver-rep3_R2.fq.gz
</code></pre>
<p>As recommended above, if all of your raw data are located in <strong>rawdata</strong>, then create a <code>rawdata/samples.xls</code> file like:</p>
<pre><code>lung-rep1
lung-rep2
lung-rep3
liver.rep1
liver.rep2
liver.rep3
</code></pre>
<p><strong>Note</strong>: You cannot mix Paired-end and Single-end samples within the same NGSPipe run as this will cause an ERROR. NGSPipe only support Paired-end samples.</p>
<h4 id="reference-data">reference data</h4>
<p>You can download reference data from NCBI, Ensembl, or anywhere else to <code>Genomedata</code>. The most import file is genome in Fasta formant and gene annotation in GFF/GTF format.</p>
<pre><code>testdata/
├── GRCm38.83.chr19.gtf
└── chr19.fa
</code></pre>
<h4 id="edit-config-file">edit config file</h4>
<p>In this section, you will need to specify every term to match your own machine, reference genome, and data.</p>
<pre><code class="language-yaml"># path relative to where you run snakemake

## reference ##
genomeAnno: &quot;testdata/GRCm38.83.chr19.gtf&quot; # gene annotation file, can be gtf or gff
genomeFasta: &quot;testdata/chr19.fa&quot; # genome sequence

## output directory ##
resultsDir: &quot;results/result&quot;
reportsDir: &quot;results/report&quot;
dbDir: &quot;results/sqlite3&quot;

## input ##
samplesList: &quot;testdata/samples.xls&quot; # sample file
samplesDir: &quot;testdata&quot; # sample file directory
read1Suffix: &quot;_R1.fq.gz&quot; # fastq suffix, read1
read2Suffix: &quot;_R2.fq.gz&quot;
replict_num: 3 # replict can by 1,2,3
# condition for differential expression by deseq2
</code></pre>
<h4 id="condition-for-sample-compare">condition for sample compare</h4>
<p>To compare each sample, a group information is needed.</p>
<pre><code>sample_id,Sample,Tissue
control-0,control,normal
control-1,control,normal
control-2,control,normal
treated-0,treated,tumor
treated-1,treated,tumor
treated-2,treated,tumor
</code></pre>
<h4 id="edit-snakefile">edit snakefile</h4>
<p>edit file <code>ngspipe/workflow/rnaseq_analysis.Snakefile.py</code> for advance setting, such as sampling data method, mapping tool, and email address to receive run log.</p>
<pre><code class="language-python"># relative path
snake_dir = workflow.basedir # all configfile, scripts, restructuretext, ens are relative to snakefile (this file)
working_dir = os.getcwd() # input and output path are relative to current working directory

# get notice
receiver_email = 'youemailaddress'

# ----------------------------------------------------------------------- #
# sample information #
#
smpList = pd.read_csv(config[&quot;samplesList&quot;], index_col=0, header=None)
SAMPLES = list(smpList.index)[0:] # how many sample you want to run.
# ----------------------------------------------------------------------- #


# -------------------------------------------------------------------------------------------------------------------------------------------------------------------- #
# detail parameters in pipe #
#
# 1. sampling data
# for test the pipe, you can choose to the part of the input file, can be whole,head:40000,tail:40000,random:0.5,random:40000
sampling_method = 'links' # tail, seqkit_number, seqkit_proportion, head, tail
sampling_data_outdir = join(config[&quot;resultsDir&quot;], &quot;sampling_data&quot;, &quot;sampling_data_by_{}&quot;.format(sampling_method))

# 2. raw reads qc
qc_method = 'trim-galore' # trimomatic
qc_outdir = join(config[&quot;resultsDir&quot;], &quot;rawReads_qc&quot;, &quot;rawReads_qc_by_{}&quot;.format(qc_method))

# 3. junction alignmnet
junction_align_method = 'hisat2' # star
junction_align_outdir = join(config[&quot;resultsDir&quot;], &quot;junction_align&quot;, &quot;junction_align_by_{}&quot;.format(junction_align_method))
genome_index_prefix = &quot;genome&quot;
# rna-seq sequencing type, can be fr-firststrand, none, fr-secondstrand
rna_library = &quot;&quot; # &quot;--rna-strandness RF&quot;(fr-firststrand) or &quot;--rna-strandness FR&quot;(fr-secondstrand)

# 4. transcript assembly
transcript_assembly_method = 'stringtie' # star
transcript_assembly_outdir = join(config[&quot;resultsDir&quot;], &quot;transcript_assembly&quot;, &quot;transcript_assembly_by_{}&quot;.format(transcript_assembly_method))

# 5. quantification
quantify_method = 'stringtie' # htseqcounts or featurecounts
quantify_outdir = join(config[&quot;resultsDir&quot;], &quot;quantify&quot;, &quot;quantify_by_{}&quot;.format(quantify_method))

# 6. statistic
statistic_data_all = [
                  '0.genomeFa', 
                  '0.genomeAnno', 
                  '1.rawReads', 
                  '2.cleanReads', 
                  '2.multiqc', 
                  '3.bam', 
                  '4.mergedGtf', 
                  '5.exp',
                  ]
genomeFa_outdir, genomeAnno_outdir, rawReads_outdir, cleanReads_outdir, multiqc_outdir, bam_outdir, mergedGtf_outdir, exp_outdir \
  = [join(config[&quot;resultsDir&quot;], &quot;statistic&quot;, &quot;statistic_data_of_{}&quot;.format(i)) for i in statistic_data_all]

statistic_data_choose = [
                  #'0.genomeFa', 
                  #'0.genomeAnno', 
                  '1.rawReads', 
                  '2.cleanReads', 
                  #'2.multiqc', 
                  '3.bam', 
                  #'4.mergedGtf', 
                  #'5.exp',
                  ]
stat_outdir = join(config[&quot;resultsDir&quot;], &quot;statistic&quot;)

# 7. diff gene discovery
diff_outdir = join(config[&quot;resultsDir&quot;], &quot;diff&quot;, &quot;diff_by_{}&quot;.format('deseq2'))
# -------------------------------------------------------------------------------------------------------------------------------------------------------------------- #

# ------------------------------
#report: &quot;report/workflow.rst&quot;
report_outdir = join(config[&quot;reportsDir&quot;], &quot;report.html&quot;)
# ------------------------------

# include modules
include: join(&quot;rules&quot;, &quot;1.sampling_data_by_{}.Snakefile.py&quot;.format(sampling_method))
include: join(&quot;rules&quot;, &quot;2.rawReads_qc_by_{}.Snakefile.py&quot;.format(qc_method))
include: join(&quot;rules&quot;, &quot;3.junction_align_by_{}.Snakefile.py&quot;.format(junction_align_method))
include: join(&quot;rules&quot;, &quot;4.transcript_assembly_by_{}.Snakefile.py&quot;.format(transcript_assembly_method))
include: join(&quot;rules&quot;, &quot;5.quant_by_{}.Snakefile.py&quot;.format(quantify_method))
include: join(&quot;rules&quot;, &quot;6.statistic_data_of_bam.Snakefile.py&quot;)
include: join(&quot;rules&quot;, &quot;6.statistic_data_of_rawReads.Snakefile.py&quot;)
include: join(&quot;rules&quot;, &quot;6.statistic_data_of_cleanReads.Snakefile.py&quot;)
include: join(&quot;rules&quot;, &quot;7.report.Snakefile.py&quot;)
include: join(&quot;rules&quot;, &quot;9.differential_expression.deseq2.Snakefile.py&quot;)

</code></pre>
<p><strong>Note</strong>: The script/env/include path is always relative to the Snakefile containing the directive (in contrast to the input, output and log file paths, which are relative to the working directory). </p>
<h4 id="report">report</h4>
<p>edit <code>ngspipe/report/*.rst</code> that will be added at the end of the report. For example, edit the <code>ngspipe/report/rawreads_stat.rst</code> file to include a text describing the 'Statistic' of the experiment. This text will be added to the report as static section and is one of the two report sections that can be edited by the end-user.</p>
<p>NGSPipe will provide you with an interactive, browser-based report, showing the most important measures of your data on the first sight. All tables in the report can be sorted and filtered. The table on the first tab shows the key values for a quick estimation of the success of your sequencing experiment and the assembly. On the second tab, there is a more detailed table, giving many additional measures. Additionally to the tables, many measures are provided as graphical feedback. On the third tab, you see plots which are generated for one complete sequencing experiment. On the fourth tab, there are plots which each show measures on one specific dataset.</p>
<pre><code class="language-matlab">
summary of reads produced
=========================

The details of the data quality are as follows:

#. Sample name: Sample name
#. Raw reads: Count raw sequence data, with four rows in a unit, count the number of sequencing sequences in each file
#. Clean reads: The calculation method is the same as that of Raw reads, except that the statistical files are filtered sequencing data, and subsequent bioinformatics analysis is based on Clean reads
#. Clean bases: Multiply the number of sequencing sequences by the length of the sequencing sequence and convert it to G as the unit
#. Error rate: base sequencing error rate
#. Q20, Q30: respectively represent the percentage of bases whose Phred value is greater than 20 and 30 to the total bases
#. GC content: the percentage of the total number of bases G and C to the total number of bases
</code></pre>
<h4 id="run-snakemake">run snakemake</h4>
<pre><code class="language-shell"># dry run, use -n parameter only print task plan, -p print commands
snakemake -s ngspipe/workflow/rnaseq_analysis.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -np

# run pipe
snakemake -s ngspipe/workflow/rnaseq_analysis.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -p -j &lt;thread number&gt;

# after run pipe successfully, report can be generated.
snakemake --snakefile ngspipe/workflow/rnaseq_analysis.Snakefile.py --report results/Report/report.html
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../NGSPipe-ChIP-seq/" class="btn btn-neutral float-right" title="NGSPipe ChIP-seq">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../NGSPipeDb/" class="btn btn-neutral" title="NGSPipeDb Tutorial"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
      <p>Bioinformatics Group, XTBG, 2021</p>
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/xuanblo/NGSPipeDb/mkdocs/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../NGSPipeDb/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../NGSPipe-ChIP-seq/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
