
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="NGSPipeDb is a pipeline">
      
      
      
        <meta name="author" content="Xuan Zhang">
      
      
        <link rel="canonical" href="https://xuanblo.github.io/NGSPipeDb/NGSPipe-RNA-seq/">
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.0.6">
    
    
      
        <title>NGSPipe RNA-seq - NGSPipeDb Document</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2c0c5eaf.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.7fa14f5b.min.css">
        
          
          
          <meta name="theme-color" content="#7e56c2">
        
      
    
    

<script src="../assets/pymdownx-extras/material-extra-theme-7c147bb7.js" type="text/javascript"></script>

    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../assets/pymdownx-extras/extra-25cd689912.css">
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="dracula" data-md-color-primary="deep-purple" data-md-color-accent="deep-purple">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#rna-seq-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="Header"
  >
    <a
      href="https://xuanblo.github.io/NGSPipeDb/"
      title="NGSPipeDb Document"
      class="md-header__button md-logo"
      aria-label="NGSPipeDb Document"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGSPipeDb Document
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              NGSPipe RNA-seq
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7l2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10l1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10l-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17.75 4.09l-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91l-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5l1.43 1.43L4.93 20.5 3.5 19.07 19.07 3.5m-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13.51 1.67m-3.3 3.61l-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87.4 1.31M19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58l2.77-1.15-.24 3.35-2.53-2.2m4.33-2.7l1.15-2.77 2.2 2.54-3.35.23m1.15-4.96l-1.14-2.78 3.34.24-2.2 2.54M9.63 18.93l2.77 1.15-2.53 2.19-.24-3.34z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M15.07 11.25l-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10z"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/xuanblo/NGSPipeDb/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="./" class="md-tabs__link md-tabs__link--active">
        Pipelines
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../linux/" class="md-tabs__link">
        Learning materials
      </a>
    </li>
  

      
        
  
  


  <li class="md-tabs__item">
    <a href="../changelog/" class="md-tabs__link">
      Change logs
    </a>
  </li>

      
        
  
  


  <li class="md-tabs__item">
    <a href="../todo/" class="md-tabs__link">
      Todo
    </a>
  </li>

      
    </ul>
  </div>
</nav>
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="NGSPipeDb Document" class="md-nav__button md-logo" aria-label="NGSPipeDb Document" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    NGSPipeDb Document
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/xuanblo/NGSPipeDb/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_1" type="checkbox" id="__nav_1" >
      
      <label class="md-nav__link" for="__nav_1">
        Home
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Home" data-md-level="1">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Getting start
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSPipeDb/" class="md-nav__link">
        NGSPipeDb Tutorial
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" checked>
      
      <label class="md-nav__link" for="__nav_2">
        Pipelines
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pipelines" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Pipelines
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          NGSPipe RNA-seq
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        NGSPipe RNA-seq
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#quick-start-one-time-installation-of-components-necessary-for-rna-seq-analysis" class="md-nav__link">
    Quick Start - One time installation of components necessary for RNA-Seq analysis 
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-by-step-rna-seq-workflow-on-testdata" class="md-nav__link">
    Step-by-step RNA-seq workflow on testdata 
  </a>
  
    <nav class="md-nav" aria-label="Step-by-step RNA-seq workflow on testdata ">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-install-wget-and-git" class="md-nav__link">
    1. Install wget and git 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-install-miniconda3" class="md-nav__link">
    2. Install Miniconda3 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-download-ngspipedb-source-code" class="md-nav__link">
    3. Download NGSPipeDb source code 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-choose-a-good-project-name" class="md-nav__link">
    4. Choose a good project name
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-installing-the-ngspipe-rna-seq-conda-environments" class="md-nav__link">
    5. Installing the NGSPipe RNA-Seq conda environments 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-download-test-files" class="md-nav__link">
    6. Download test files 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-run-rna-seq-analysis-on-test-data" class="md-nav__link">
    7. run RNA-seq analysis on test data 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#8-rgenerate-report" class="md-nav__link">
    8. rgenerate report 
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-your-custome-data" class="md-nav__link">
    run your custome data 
  </a>
  
    <nav class="md-nav" aria-label="run your custome data ">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-rawdata-sequence-data" class="md-nav__link">
    1. Rawdata sequence data 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-reference-data" class="md-nav__link">
    2. Reference data 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-edit-config-file" class="md-nav__link">
    3. Edit config file 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-condition-for-sample-compare" class="md-nav__link">
    4. Condition for sample compare 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-edit-snakefile" class="md-nav__link">
    5. edit snakefile 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-custome-report" class="md-nav__link">
    6. Custome report 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-run-snakemake" class="md-nav__link">
    7. Run snakemake 
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSPipe-ChIP-seq/" class="md-nav__link">
        NGSPipe ChIP-seq
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSPipe-resequecing/" class="md-nav__link">
        NGSPipe resequecing
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSDb/" class="md-nav__link">
        NGSDb
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      <label class="md-nav__link" for="__nav_3">
        Learning materials
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Learning materials" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Learning materials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../linux/" class="md-nav__link">
        Linux & shell basic
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../ngs/" class="md-nav__link">
        NGS analysis
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../conda/" class="md-nav__link">
        Conda
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        Change logs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../todo/" class="md-nav__link">
        Todo
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#quick-start-one-time-installation-of-components-necessary-for-rna-seq-analysis" class="md-nav__link">
    Quick Start - One time installation of components necessary for RNA-Seq analysis 
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-by-step-rna-seq-workflow-on-testdata" class="md-nav__link">
    Step-by-step RNA-seq workflow on testdata 
  </a>
  
    <nav class="md-nav" aria-label="Step-by-step RNA-seq workflow on testdata ">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-install-wget-and-git" class="md-nav__link">
    1. Install wget and git 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-install-miniconda3" class="md-nav__link">
    2. Install Miniconda3 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-download-ngspipedb-source-code" class="md-nav__link">
    3. Download NGSPipeDb source code 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-choose-a-good-project-name" class="md-nav__link">
    4. Choose a good project name
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-installing-the-ngspipe-rna-seq-conda-environments" class="md-nav__link">
    5. Installing the NGSPipe RNA-Seq conda environments 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-download-test-files" class="md-nav__link">
    6. Download test files 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-run-rna-seq-analysis-on-test-data" class="md-nav__link">
    7. run RNA-seq analysis on test data 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#8-rgenerate-report" class="md-nav__link">
    8. rgenerate report 
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-your-custome-data" class="md-nav__link">
    run your custome data 
  </a>
  
    <nav class="md-nav" aria-label="run your custome data ">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-rawdata-sequence-data" class="md-nav__link">
    1. Rawdata sequence data 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-reference-data" class="md-nav__link">
    2. Reference data 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-edit-config-file" class="md-nav__link">
    3. Edit config file 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-condition-for-sample-compare" class="md-nav__link">
    4. Condition for sample compare 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-edit-snakefile" class="md-nav__link">
    5. edit snakefile 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-custome-report" class="md-nav__link">
    6. Custome report 
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-run-snakemake" class="md-nav__link">
    7. Run snakemake 
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
  
                
                  <a href="https://github.com/xuanblo/NGSPipeDb/edit/master/mkdocs/markdowns/NGSPipe-RNA-seq.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="rna-seq-analysis">RNA-seq analysis</h1>
<h2 id="quick-start-one-time-installation-of-components-necessary-for-rna-seq-analysis">Quick Start - One time installation of components necessary for RNA-Seq analysis <a name="QuickStarted"></a></h2>
<div class="tabbed-set" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">Linux &amp; WSL</label><div class="tabbed-content">
<p>1.download ngspipedb to anywhere you want
<div class="highlight"><pre><span></span><code>git clone git://github.com/xuanblo/NGSPipeDb.git
mv NGSPipeDb mouse_transcriptome_analysis
<span class="nb">cd</span> mouse_transcriptome_analysis
</code></pre></div></p>
<p>2.create conda environment, download test data, and run RNA-Seq analysis
<div class="highlight"><pre><span></span><code>bash ngspipe/scripts/one_step_ranseq_test.sh linux
</code></pre></div></p>
</div>
<input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><label for="__tabbed_1_2">MacOSX</label><div class="tabbed-content">
<p>1.download ngspipedb to anywhere you want
<div class="highlight"><pre><span></span><code>git clone git://github.com/xuanblo/NGSPipeDb.git
mv NGSPipeDb mouse_transcriptome_analysis
<span class="nb">cd</span> mouse_transcriptome_analysis
</code></pre></div></p>
<p>2.create conda environment, download test data, and run RNA-Seq analysis
<div class="highlight"><pre><span></span><code>bash ngspipe/scripts/one_step_ranseq_test.sh macos
</code></pre></div></p>
</div>
</div>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>If you have more time, then we recommend you configure ngspipedb according to your needs. For more details, please see <a href="#step-by-step">step by step</a> bellow.</p>
</div>
<h2 id="step-by-step-rna-seq-workflow-on-testdata">Step-by-step RNA-seq workflow on testdata <a href="Step-by-Step-RNASeq"></a></h2>
<p>Although included in this section are step-by-step instructions, it is assumed that the user has a basic understanding of the <a href="https://en.wikipedia.org/wiki/Command-line_interface">nix command line interface</a>. Also, it would be better if the user has basic knowledge about <a href="">snakemake</a>, <a href="">conda</a> and <a href="">best practice RNA sequence analysis</a>, but it is not required. You can find some easy-to-learn matierals in <a href="../linux">linux &amp; shell</a> and <a href="../ngs#rnaseq">RNASeq background</a>.</p>
<h3 id="1-install-wget-and-git">1. Install wget and git <a name="BasicLinux"></a></h3>
<p>To get some of the required software packages, we will use the command line tools called <a href="http://www.gnu.org/software/wget/">wget</a> and <a href="https://git-scm.com/book/en/v2/Getting-Started-Installing-Git">git</a>.  <em>wget</em> is a popular tool for downloading things off of the internet.  <em>git</em> is a distributed version control system which we will use to checkout the NGSPipeDb code.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These tools are already pre-installed in most systems, but if you are unsure whether or not you have <em>wget</em> enter <code>wget</code> and if the return is <code>wget: command not found</code>, then you will have to install <em>wget</em>.  Do likewise for <em>git</em>.</p>
</div>
<h3 id="2-install-miniconda3">2. Install Miniconda3 <a name="Miniconda"></a></h3>
<p>NGSPipeDb relies on the conda package manager for installation and dependency resolution, so you will need to install <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html">conda</a> first.</p>
<p>We will be using the <a href="http://conda.pydata.org/miniconda.html">Miniconda3</a> package management system (aka <strong>CONDA</strong>) to manage all of the software packages that <strong>NGSPipeDb</strong> is dependent on. </p>
<p>Use following commands to retrieve and then run the Minicoda3 installation script:</p>
<p>1.download miniconda3
<div class="highlight"><pre><span></span><code>wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
</code></pre></div></p>
<p>2.install miniconda3
<div class="highlight"><pre><span></span><code>bash Miniconda3-latest-Linux-x86_64.sh
</code></pre></div></p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>While running the installation script, follow the commands listed on screen, and press the <em>enter</em> key to scroll. Make sure to answer <code>yes</code> when asked if you want to prepend Miniconda3 to PATH. After that, close your terminal, open a new one and you should now have Conda working! You could, alternatively, run <code class="highlight"><span class="nb">source</span> ~/.bashrc</code> to initiate conda.  </p>
</div>
<p>3.Test by entering: <code>conda update conda</code>. Press <code>y</code> to confirm the conda updates.  </p>
<p>4.Finally, <code>conda install mamba -c conda-forge</code>. Mamba is a reimplementation of the conda package manager in C++, the fast conda-alternative.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You will only have to install Minicoda3 once.</p>
</div>
<h3 id="3-download-ngspipedb-source-code">3. Download NGSPipeDb source code <a name="NGSPipeDbSource"></a></h3>
<p>We offer two download options.</p>
<div class="tabbed-set" data-tabs="2:2"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">latest stable version</label><div class="tabbed-content">
<p>To install the latest stable version of NGSPipeDb, please clone the git repository to your system. First, <code>cd /path/to/where_you_want</code>, and then
<div class="highlight"><pre><span></span><code>git clone https://www.github.com/xuanblo/NGSPipeDb
</code></pre></div></p>
<div class="admonition error">
<p class="admonition-title">Error</p>
<p>If you meet problem with network, please try to download the "specific version" or modify the command to:<br />
<code>git clone git://www.github.com/xuanblo/NGSPipeDb</code>.</p>
</div>
</div>
<input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><label for="__tabbed_2_2">specific version</label><div class="tabbed-content">
<p>If you want to use specific version, please have a look at the <a href="../changelog">change log</a> and different <a href="https://github.com/xuanblo/NGSPipeDb/releases">release version</a> first.</p>
<p>Then, <code class="highlight"><span class="nb">cd</span> /path/to/where_you_want</code>
<div class="highlight"><pre><span></span><code>wget https://github.com/xuanblo/NGSPipeDb/archive/NGSPipeDb_v0.0.1.tar.gz
tar -xf NGSPipeDb_v0.0.1.tar.gz
</code></pre></div></p>
</div>
</div>
<h3 id="4-choose-a-good-project-name">4. Choose a good project name</h3>
<p>Because all analysis will be done directly in the source code directory. So you can change the directory name in the following way.</p>
<div class="tabbed-set" data-tabs="3:2"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">download by git</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code>mv NGSPipeDb species_sample_transcript_analysis_by_NGSPipeDb
</code></pre></div>
</div>
<input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><label for="__tabbed_3_2">download by wget</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code>mv NGSPipeDb-NGSPipeDb_v0.0.1 species_sample_transcript_analysis_by_NGSPipeDb
</code></pre></div>
</div>
</div>
<p>Enter directory <code class="highlight"><span class="nb">cd</span> species_sample_transcript_analysis_by_NGSPipeDb</code> for further setps.</p>
<p>You can view the <strong>NGSPipe</strong> code structure with <code class="highlight">tree ./ngspipe/ -d -L <span class="m">2</span></code></p>
<div class="highlight"><pre><span></span><code>./ngspipe/
├── config
├── envs
├── imgs
├── metadata
├── notebooks
├── reports
├── rules
└── scripts
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Files other than <code>ngspipe</code> are not required for RNA-Seq analysis.</p>
</div>
<h3 id="5-installing-the-ngspipe-rna-seq-conda-environments">5. Installing the NGSPipe RNA-Seq conda environments <a name="NGSPipeDbEnv"></a></h3>
<p>First, you will need to create a conda environments. For details, see <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html">manage envirement</a> with conda.</p>
<div class="highlight"><pre><span></span><code>conda create -c conda-forge -c bioconda --name ngspipe-rnaseq <span class="nv">snakemake</span><span class="o">=</span><span class="m">5</span>.30.2 <span class="nv">python</span><span class="o">=</span><span class="m">3</span>.8 <span class="nv">seqkit</span><span class="o">=</span><span class="m">0</span>.14.0
</code></pre></div>
<p>Next, to producess and analysis the RNA-Seq sequenceding data, some bioinformatics tools need to be installed. Chick to <a href="https://github.com/xuanblo/NGSPipeDb/blob/master/ngspipe/envs/requirements_rnaseq.yaml">see software we used</a>.</p>
<div class="highlight"><pre><span></span><code>conda env update -n ngspipe-rnaseq --file ngspipe/envs/requirements_rnaseq.yaml --prune
</code></pre></div>
<p>Activate conda environment:</p>
<div class="highlight"><pre><span></span><code>conda activate ngspipe-rnaseq
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>By default, all softwares and packages are stored under <code>~/miniconda/conda/env/ngspipe-rnaseq</code>.</p>
</div>
<div class="admonition error">
<p class="admonition-title">Error</p>
<p>If the conda downloading encounters any problems, you can refer to how to <a href="../conda/#pre-build">download pre-build ngspipe-ranseq env</a>.</p>
</div>
<h3 id="6-download-test-files">6. Download test files <a name="Testdata"></a></h3>
<p><strong>NGSPipe</strong> is dependent on reference files and raw sequence reads which can be found in <a href="http://www.liu-lab.com/ngspipedb/testdata">http://www.liu-lab.com/ngspipedb/testdata</a></p>
<p>To download the mouse RNA-seq test data into <code>./testdata</code>, you can choose auto download with a script or Manually:</p>
<div class="tabbed-set" data-tabs="4:2"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">auto download by a shell script</label><div class="tabbed-content">
<div class="highlight"><pre><span></span><code>bash ngspipe/scripts/download_testdata.sh testdata
</code></pre></div>
</div>
<input id="__tabbed_4_2" name="__tabbed_4" type="radio" /><label for="__tabbed_4_2">Manually download by wget</label><div class="tabbed-content">
<p>1.download the test files (160M) that you need and then untarring then in a directory called <code>testdata</code>.
<div class="highlight"><pre><span></span><code>wget http://www.liu-lab.com/ngspipedb/rnaseq_testdata.tar.gz
tar -zxvf rnaseq_testdata.tar.gz
</code></pre></div></p>
<ol>
<li>generate replicated samples:
<div class="highlight"><pre><span></span><code><span class="nb">cd</span> testdata
python ../ngspipe/scripts/generate_replicat.py control_R1.fq.gz control_R2.fq.gz treated_R1.fq.gz treated_R2.fq.gz
gunzip chr19.fa.gz
gunzip GRCm38.83.chr19.gtf.gz
<span class="nb">cd</span> ..
</code></pre></div></li>
</ol>
</div>
</div>
<p>Make sure you have the following directory structure by command <code>tree testdata</code>:</p>
<div class="highlight"><pre><span></span><code>testdata
├── chr19.fa
├── condition.xls
├── control-0_R1.fq.gz
├── control-0_R2.fq.gz
├── control-1_R1.fq.gz
├── control-1_R2.fq.gz
├── control-2_R1.fq.gz
├── control-2_R2.fq.gz
├── control_R1.fq.gz
├── control_R2.fq.gz
├── GRCm38.83.chr19.gtf
├── samples.xls
├── treated-0_R1.fq.gz
├── treated-0_R2.fq.gz
├── treated-1_R1.fq.gz
├── treated-1_R2.fq.gz
├── treated-2_R1.fq.gz
├── treated-2_R2.fq.gz
├── treated_R1.fq.gz
└── treated_R2.fq.gz
</code></pre></div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The test data is only used to verify that the analytical process is working properly and the results do not have r biological significance.</p>
</div>
<h3 id="7-run-rna-seq-analysis-on-test-data">7. run RNA-seq analysis on test data <a name="RunTest"></a></h3>
<p>We provied a simple RNA-seq workflow for you to take a glance of NGSPipe. In RNA-seq analysis part, it contains 7 step analysis:</p>
<div class="highlight"><pre><span></span><code><span class="mf">1.</span> <span class="n">sampling</span> <span class="n">data</span> <span class="p">(</span><span class="n">choose</span> <span class="n">part</span> <span class="n">of</span> <span class="n">your</span> <span class="n">data</span><span class="p">)</span>
<span class="mf">2.</span> <span class="n">raw</span> <span class="n">reads</span> <span class="n">qc</span>
<span class="mf">3.</span> <span class="n">junction</span> <span class="n">align</span> <span class="n">to</span> <span class="n">genome</span>
<span class="mf">4.</span> <span class="n">transcript</span> <span class="n">assembly</span>
<span class="mf">5.</span> <span class="n">gene</span> <span class="n">quantification</span>
<span class="mf">6.</span> <span class="n">statistic</span>
<span class="mf">7.</span> <span class="n">differential</span> <span class="n">gene</span> <span class="n">analysis</span>
</code></pre></div>
<p>First, we need to check the <code>ngspipe/file</code>. This will not execute anything, but display what would be done.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># dry run, use -n parameter only print task plan, -p print commands</span>
snakemake -s ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -np
</code></pre></div>
<p>If nothing goes wrong, you can generate a dag plot:</p>
<div class="highlight"><pre><span></span><code>snakemake -s ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml --dag<span class="p">|</span>dot -Tpng &gt; dag.png
</code></pre></div>
<p>Now you can do RNA-seq analysis by just one simply command.</p>
<div class="highlight"><pre><span></span><code>snakemake -s ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -p -j <span class="m">10</span>
</code></pre></div>
<p>The final data files are put in the folder <code>results</code>. Please check you result file <code>tree -d -L 2 results/</code>, it may like this:</p>
<div class="highlight"><pre><span></span><code>results/rnaseq
├── report
│   ├── 1.rawreads_stat
│   ├── 2.cleanreads_stat
│   ├── 3.mapping_stat
│   ├── 4.exp_stat
│   └── workflow
└── result
    ├── diff
    ├── junction_align
    ├── quantify
    ├── rawReads_qc
    ├── sampling_data
    └── statistic
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you encounter any problem in this step, please turn to <code>TroubleShooting</code> for help.</p>
</div>
<h3 id="8-rgenerate-report">8. rgenerate report <a name="Report"></a></h3>
<p>If all goes well, the proper analysis will be followed by the making of the html report using Snakemake to a html report file with pictures and tables.</p>
<div class="highlight"><pre><span></span><code>snakemake --snakefile ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml --report results/report/report.html
</code></pre></div>
<p>The final report should appear as <code>results/report/report.html</code>. This report is a single html file with all in it and can be sent to customers/colleagues as a final report. It is nicer than a PDF version because of large tables and figures which would suffer from page breaks and it can be viewed on any device supporting html include smartphones :-).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Internet Explorer is not supported.</p>
</div>
<h2 id="run-your-custome-data">run your custome data <a name="RunRawdata"></a></h2>
<p>NGSPipe is built to be used routinely. To ensure a maximum comparability of the results, you can copy default <code>ngspipe/config/rnaseq.config.yaml</code> and <code>ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py</code> file to the same directory:</p>
<p><div class="highlight"><pre><span></span><code>cp ngspipe/config/rnaseq.config.yaml ngspipe/config/my_own_rnaseq.config.yaml 
cp ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py ngspipe/my_own_rnaseq_analysis.Snakefile.py
</code></pre></div>
We will explain how to edit and configure these files shortly below.</p>
<h3 id="1-rawdata-sequence-data">1. Rawdata sequence data <a name="download_raw_data"></a></h3>
<p>Raw data files can either be fastq, fastq.gz formated files. <code>mkdir rawdata</code> and upload your own data to this directory. If your raw data are located in <strong>somewhere else</strong>, you can copy them to <code>rawdata</code>, or create soft links like <code>ln -s ../yoursamplepath/*.fq.gz rawdata/</code>.</p>
<div class="highlight"><pre><span></span><code>rawdata/
├── lung-rep1_R1.fq.gz
├── lung-rep1_R2.fq.gz
├── lung-rep2_R1.fq.gz
├── lung-rep2_R2.fq.gz
├── lung-rep3_R1.fq.gz
├── lung-rep3_R2.fq.gz
├── liver-rep1_R1.fq.gz
├── liver-rep1_R2.fq.gz
├── liver-rep2_R1.fq.gz
├── liver-rep2_R2.fq.gz
├── liver-rep3_R1.fq.gz
├── liver-rep3_R2.fq.gz
</code></pre></div>
<p>As recommended above, if all of your raw data are located in <strong>rawdata</strong>, then create a <code>rawdata/samples.xls</code> file like:</p>
<div class="highlight"><pre><span></span><code>lung-rep1
lung-rep2
lung-rep3
liver-rep1
liver-rep2
liver-rep3
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You cannot mix Paired-end and Single-end samples within the same NGSPipe run as this will cause an ERROR. NGSPipe only support Paired-end samples.</p>
</div>
<h3 id="2-reference-data">2. Reference data <a name="download_ref_data"></a></h3>
<p>You can download reference data from NCBI, Ensembl, or anywhere else to <code>genomedata</code>. The most import file is genome in Fasta formant and gene annotation in GFF/GTF format. Use the same method as rawdata does:</p>
<div class="highlight"><pre><span></span><code>genomedata/
├── GRCm38.83.chr19.gtf
└── chr19.fa
</code></pre></div>
<h3 id="3-edit-config-file">3. Edit config file <a name="edit_config"></a></h3>
<p>In this section, you will need to specify every term to match your own machine, reference genome, and data.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># path relative to where you run snakemake</span>

<span class="c1">## reference ##</span>
<span class="nt">genomeAnno</span><span class="p">:</span> <span class="s">&quot;testdata/GRCm38.83.chr19.gtf&quot;</span> <span class="c1"># gene annotation file, can be gtf or gff</span>
<span class="nt">genomeFasta</span><span class="p">:</span> <span class="s">&quot;testdata/chr19.fa&quot;</span> <span class="c1"># genome sequence</span>

<span class="c1">## output directory ##</span>
<span class="nt">resultsDir</span><span class="p">:</span> <span class="s">&quot;results/rnaseq&quot;</span>
<span class="nt">reportsDir</span><span class="p">:</span> <span class="s">&quot;results/report&quot;</span>
<span class="nt">dbDir</span><span class="p">:</span> <span class="s">&quot;results/sqlite3&quot;</span>

<span class="c1">## input ##</span>
<span class="nt">samplesList</span><span class="p">:</span> <span class="s">&quot;testdata/samples.xls&quot;</span> <span class="c1"># sample file</span>
<span class="nt">samplesDir</span><span class="p">:</span> <span class="s">&quot;testdata&quot;</span> <span class="c1"># sample file directory</span>
<span class="nt">read1Suffix</span><span class="p">:</span> <span class="s">&quot;_R1.fq.gz&quot;</span> <span class="c1"># fastq suffix, read1</span>
<span class="nt">read2Suffix</span><span class="p">:</span> <span class="s">&quot;_R2.fq.gz&quot;</span>
<span class="nt">replict_num</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">3</span> <span class="c1"># replict can by 1,2,3</span>
<span class="c1"># condition for differential expression by deseq2</span>
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The input, output file paths are relative to the working directory. </p>
</div>
<h3 id="4-condition-for-sample-compare">4. Condition for sample compare <a name="edit_condition"></a></h3>
<p>To perform gene differential analysis, please create a <code>rawdata/condition.xls</code> file.</p>
<div class="highlight"><pre><span></span><code>sample_id,Sample,Tissue
lung-rep1,lung,normal
lung-rep2,lung,normal
lung-rep3,lung,normal
liver-rep1,liver,tumor
liver-rep2,liver,tumor
liver-rep3,liver,tumor
</code></pre></div>
<h3 id="5-edit-snakefile">5. edit snakefile <a name="edit_snakefile"></a></h3>
<p>Edit file <code>ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py</code> for advance setting, such as sampling data method, mapping tool, and email address to receive run log.</p>
<div class="highlight"><pre><span></span><code><span class="c1"># relative path</span>
<span class="n">snake_dir</span> <span class="o">=</span> <span class="n">workflow</span><span class="o">.</span><span class="n">basedir</span> <span class="c1"># all configfile, scripts, restructuretext, ens are relative to snakefile (this file)</span>
<span class="n">working_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span> <span class="c1"># input and output path are relative to current working directory</span>

<span class="c1"># get notice</span>
<span class="n">receiver_email</span> <span class="o">=</span> <span class="s1">&#39;youemailaddress&#39;</span>

<span class="c1"># ----------------------------------------------------------------------- #</span>
<span class="c1"># sample information #</span>
<span class="c1">#</span>
<span class="n">smpList</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;samplesList&quot;</span><span class="p">],</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">SAMPLES</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">smpList</span><span class="o">.</span><span class="n">index</span><span class="p">)[</span><span class="mi">0</span><span class="p">:]</span> <span class="c1"># how many sample you want to run.</span>
<span class="c1"># ----------------------------------------------------------------------- #</span>


<span class="c1"># -------------------------------------------------------------------------------------------------------------------------------------------------------------------- #</span>
<span class="c1"># detail parameters in pipe #</span>
<span class="c1">#</span>
<span class="c1"># 1. sampling data</span>
<span class="c1"># for test the pipe, you can choose to the part of the input file, can be whole,head:40000,tail:40000,random:0.5,random:40000</span>
<span class="n">sampling_method</span> <span class="o">=</span> <span class="s1">&#39;links&#39;</span> <span class="c1"># tail, seqkit_number, seqkit_proportion, head, tail</span>
<span class="n">sampling_data_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;sampling_data&quot;</span><span class="p">,</span> <span class="s2">&quot;sampling_data_by_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sampling_method</span><span class="p">))</span>

<span class="c1"># 2. raw reads qc</span>
<span class="n">qc_method</span> <span class="o">=</span> <span class="s1">&#39;trim-galore&#39;</span> <span class="c1"># trimomatic</span>
<span class="n">qc_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;rawReads_qc&quot;</span><span class="p">,</span> <span class="s2">&quot;rawReads_qc_by_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">qc_method</span><span class="p">))</span>

<span class="c1"># 3. junction alignmnet</span>
<span class="n">junction_align_method</span> <span class="o">=</span> <span class="s1">&#39;hisat2&#39;</span> <span class="c1"># star</span>
<span class="n">junction_align_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;junction_align&quot;</span><span class="p">,</span> <span class="s2">&quot;junction_align_by_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">junction_align_method</span><span class="p">))</span>
<span class="n">genome_index_prefix</span> <span class="o">=</span> <span class="s2">&quot;genome&quot;</span>
<span class="c1"># rna-seq sequencing type, can be fr-firststrand, none, fr-secondstrand</span>
<span class="n">rna_library</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># &quot;--rna-strandness RF&quot;(fr-firststrand) or &quot;--rna-strandness FR&quot;(fr-secondstrand)</span>

<span class="c1"># 4. transcript assembly</span>
<span class="n">transcript_assembly_method</span> <span class="o">=</span> <span class="s1">&#39;stringtie&#39;</span> <span class="c1"># star</span>
<span class="n">transcript_assembly_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;transcript_assembly&quot;</span><span class="p">,</span> <span class="s2">&quot;transcript_assembly_by_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">transcript_assembly_method</span><span class="p">))</span>

<span class="c1"># 5. quantification</span>
<span class="n">quantify_method</span> <span class="o">=</span> <span class="s1">&#39;stringtie&#39;</span> <span class="c1"># htseqcounts or featurecounts</span>
<span class="n">quantify_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;quantify&quot;</span><span class="p">,</span> <span class="s2">&quot;quantify_by_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">quantify_method</span><span class="p">))</span>

<span class="c1"># 6. statistic</span>
<span class="n">statistic_data_all</span> <span class="o">=</span> <span class="p">[</span>
                  <span class="s1">&#39;0.genomeFa&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;0.genomeAnno&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;1.rawReads&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;2.cleanReads&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;2.multiqc&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;3.bam&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;4.mergedGtf&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;5.exp&#39;</span><span class="p">,</span>
                  <span class="p">]</span>
<span class="n">genomeFa_outdir</span><span class="p">,</span> <span class="n">genomeAnno_outdir</span><span class="p">,</span> <span class="n">rawReads_outdir</span><span class="p">,</span> <span class="n">cleanReads_outdir</span><span class="p">,</span> <span class="n">multiqc_outdir</span><span class="p">,</span> <span class="n">bam_outdir</span><span class="p">,</span> <span class="n">mergedGtf_outdir</span><span class="p">,</span> <span class="n">exp_outdir</span> \
  <span class="o">=</span> <span class="p">[</span><span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;statistic&quot;</span><span class="p">,</span> <span class="s2">&quot;statistic_data_of_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">statistic_data_all</span><span class="p">]</span>

<span class="n">statistic_data_choose</span> <span class="o">=</span> <span class="p">[</span>
                  <span class="c1">#&#39;0.genomeFa&#39;, </span>
                  <span class="c1">#&#39;0.genomeAnno&#39;, </span>
                  <span class="s1">&#39;1.rawReads&#39;</span><span class="p">,</span> 
                  <span class="s1">&#39;2.cleanReads&#39;</span><span class="p">,</span> 
                  <span class="c1">#&#39;2.multiqc&#39;, </span>
                  <span class="s1">&#39;3.bam&#39;</span><span class="p">,</span> 
                  <span class="c1">#&#39;4.mergedGtf&#39;, </span>
                  <span class="c1">#&#39;5.exp&#39;,</span>
                  <span class="p">]</span>
<span class="n">stat_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;statistic&quot;</span><span class="p">)</span>

<span class="c1"># 7. diff gene discovery</span>
<span class="n">diff_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;resultsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;diff&quot;</span><span class="p">,</span> <span class="s2">&quot;diff_by_</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;deseq2&#39;</span><span class="p">))</span>
<span class="c1"># -------------------------------------------------------------------------------------------------------------------------------------------------------------------- #</span>

<span class="c1"># ------------------------------</span>
<span class="c1">#report: &quot;report/workflow.rst&quot;</span>
<span class="n">report_outdir</span> <span class="o">=</span> <span class="n">join</span><span class="p">(</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;reportsDir&quot;</span><span class="p">],</span> <span class="s2">&quot;report.html&quot;</span><span class="p">)</span>
<span class="c1"># ------------------------------</span>

<span class="c1"># include modules</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;1.sampling_data_by_</span><span class="si">{}</span><span class="s2">.Snakefile.py&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sampling_method</span><span class="p">))</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;2.rawReads_qc_by_</span><span class="si">{}</span><span class="s2">.Snakefile.py&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">qc_method</span><span class="p">))</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;3.junction_align_by_</span><span class="si">{}</span><span class="s2">.Snakefile.py&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">junction_align_method</span><span class="p">))</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;4.transcript_assembly_by_</span><span class="si">{}</span><span class="s2">.Snakefile.py&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">transcript_assembly_method</span><span class="p">))</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;5.quant_by_</span><span class="si">{}</span><span class="s2">.Snakefile.py&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">quantify_method</span><span class="p">))</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;6.statistic_data_of_bam.Snakefile.py&quot;</span><span class="p">)</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;6.statistic_data_of_rawReads.Snakefile.py&quot;</span><span class="p">)</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;6.statistic_data_of_cleanReads.Snakefile.py&quot;</span><span class="p">)</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;7.report.Snakefile.py&quot;</span><span class="p">)</span>
<span class="n">include</span><span class="p">:</span> <span class="n">join</span><span class="p">(</span><span class="s2">&quot;rules&quot;</span><span class="p">,</span> <span class="s2">&quot;9.differential_expression.deseq2.Snakefile.py&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code>script/env/include</code> path is always relative to the Snakefile containing the directive (in contrast to the input, output and log file paths, which are relative to the working directory). </p>
</div>
<h3 id="6-custome-report">6. Custome report <a name="custom_report"></a></h3>
<p>edit <code>ngspipe/report/*.rst</code> that will be added at the end of the report. For example, edit the <code>ngspipe/report/rawreads_stat.rst</code> file to include a text describing the 'Statistic' of the experiment. This text will be added to the report as static section and is one of the two report sections that can be edited by the end-user.</p>
<div class="highlight"><pre><span></span><code><span class="l l-Scalar l-Scalar-Plain">summary of reads produced</span>
<span class="l l-Scalar l-Scalar-Plain">=========================</span>

<span class="l l-Scalar l-Scalar-Plain">The details of the data quality are as follows</span><span class="p p-Indicator">:</span>

<span class="c1">#. Sample name: Sample name</span>
<span class="c1">#. Raw reads: Count raw sequence data, with four rows in a unit, count the number of sequencing sequences in each file</span>
<span class="c1">#. Clean reads: The calculation method is the same as that of Raw reads, except that the statistical files are filtered sequencing data, and subsequent bioinformatics analysis is based on Clean reads</span>
<span class="c1">#. Clean bases: Multiply the number of sequencing sequences by the length of the sequencing sequence and convert it to G as the unit</span>
<span class="c1">#. Error rate: base sequencing error rate</span>
<span class="c1">#. Q20, Q30: respectively represent the percentage of bases whose Phred value is greater than 20 and 30 to the total bases</span>
<span class="c1">#. GC content: the percentage of the total number of bases G and C to the total number of bases</span>
</code></pre></div>
<p>NGSPipe will provide you with an interactive, browser-based report, showing the most important measures of your data on the first sight. All tables in the report can be sorted and filtered. The table on the first tab shows the key values for a quick estimation of the success of your sequencing experiment and the assembly. On the second tab, there is a more detailed table, giving many additional measures. Additionally to the tables, many measures are provided as graphical feedback. On the third tab, you see plots which are generated for one complete sequencing experiment. On the fourth tab, there are plots which each show measures on one specific dataset.</p>
<h3 id="7-run-snakemake">7. Run snakemake <a name="custom_run"></a></h3>
<p>Run snakemake in a prebuild environment:</p>
<div class="highlight"><pre><span></span><code>snakemake -s ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -p -j &lt;thread number&gt;
</code></pre></div>
<p>Or install dependencies while snakemake run. By default, all dependencies and tools should automatically be installed on the first execution. </p>
<div class="highlight"><pre><span></span><code>snakemake -s ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -p -j &lt;thread number&gt; --use-conda
</code></pre></div>
<p>See Specifying a location for an environment or run conda create &ndash;help for information on specifying a different path.</p>
<div class="highlight"><pre><span></span><code>snakemake -s ngspipe/1.1.rnaseq_analysis_reference_basic.Snakefile.py --configfile ngspipe/config/rnaseq.config.yaml -p -j &lt;thread number&gt; -p ./ngspipe_rnaseq_macox64
</code></pre></div>
                
                  
                
              
  <footer class="sponsorship">
    <hr>
    <a href="https://github.com/sponsors/facelessuser" title="Become a sponsor">
      <span class="twemoji heart-throb-hover">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path fill-rule="evenodd" d="M7.655 14.916L8 14.25l.345.666a.752.752 0 0 1-.69 0zm0 0L8 14.25l.345.666.002-.001.006-.003.018-.01a7.643 7.643 0 0 0 .31-.17 22.08 22.08 0 0 0 3.433-2.414C13.956 10.731 16 8.35 16 5.5 16 2.836 13.914 1 11.75 1 10.203 1 8.847 1.802 8 3.02 7.153 1.802 5.797 1 4.25 1 2.086 1 0 2.836 0 5.5c0 2.85 2.045 5.231 3.885 6.818a22.075 22.075 0 0 0 3.744 2.584l.018.01.006.003h.002z"/></svg>
      </span>
    </a>
    <hr>
  </footer>

              
                


              
            </article>
          </div>
        </div>
      </main>
      
        

<footer class="md-footer">
  
    <nav
      class="md-footer__inner md-grid"
      aria-label="Footer"
    >
      
        <a
          href="../NGSPipeDb/"
          class="md-footer__link md-footer__link--prev"
          rel="prev"
        >
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              NGSPipeDb Tutorial
            </div>
          </div>
        </a>
      
      
        <a
          href="../NGSPipe-ChIP-seq/"
          class="md-footer__link md-footer__link--next"
          rel="next"
        >
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              NGSPipe ChIP-seq
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Bioinformatics Group, XTBG, 2021
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://twitter.com/squidfunk" target="_blank" rel="noopener" title="twitter.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../assets/javascripts/workers/search.fb4a9340.min.js", "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.a1c7c35e.min.js"></script>
      
        <script src="../assets/pymdownx-extras/extra-uml-51b11a9a.js"></script>
      
    
  </body>
</html>