<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="canonical" href="https://xuanblo.github.io/NGSPipeDb/NGSDb/">
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>NGSDb - NGSPipeDb Document</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "NGSDb";
    var mkdocs_page_input_path = "NGSDb.md";
    var mkdocs_page_url = "/NGSPipeDb/NGSDb/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> NGSPipeDb Document</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Getting start</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSPipeDb/">NGSPipeDb Tutorial</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSPipe-RNA-seq/">NGSPipe RNA-seq</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSPipe-ChIP-seq/">NGSPipe ChIP-seq</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../NGSPipe-resequecing/">NGSPipe resequecing</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">NGSDb</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#one-step-to-view-the-test-database">one step to view the test database</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#step-by-step-database-generate">Step-by-step database generate </a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#1">1. 安装环境 </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#2-convert-analysis-result-to-sqlite3-file">2. convert analysis result to sqlite3 file </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#3-config">3. config </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#4-start-server">4. start server </a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#5-add-your-custome-data">5. add your custome data</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#overview-the-ngsdb-apps">overview the NGSDb apps</a>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../linux/">Linux & shell basic</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../ngs/">NGS analysis</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">NGSPipeDb Document</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>NGSDb</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/xuanblo/NGSPipeDb/mkdocs/edit/master/docs/NGSDb.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="ngsdb">NGSDb</h1>
<p><strong>Table of Contents:</strong></p>
<ol>
<li><a href="#BasicLinux">Installing wget and git</a></li>
<li><a href="#Miniconda">Installing Miniconda3</a></li>
<li><a href="#NGSPipeDbEnv">Installing the NGSPipeDb conda environments</a></li>
<li><a href="#NGSPipeDbSource">Downloading the NGSPipeDb source code</a></li>
<li><a href="#Testdata">Downloading the NGSPipeDb test files</a></li>
<li><a href="#RunTest">Run test data</a></li>
<li><a href="#Report">Generate report</a></li>
<li><a href="#RunRawdata">Run your custome data</a></li>
</ol>
<h2 id="one-step-to-view-the-test-database">one step to view the test database</h2>
<p>Now you can viste your website on http://127.0.0.1:8000. All result are stored in <code>results</code>.
- Example of report <sub><a href="http://www.liu-lab.com"><img alt="html" src="https://img.icons8.com/ios/20/000000/html-filetype.png" /></a></sub>.
- Example of database <sub><a href="http://www.liu-lab.com"><img alt="html" src="https://img.icons8.com/dotty/25/000000/copy-link.png" /></a></sub>.</p>
<h2 id="step-by-step-database-generate">Step-by-step database generate <a name="Step-by-Step-Database"></a></h2>
<h3 id="1">1. 安装环境 <a name="DatabaseRequirement"></a></h3>
<pre><code class="language-shell">conda create -n ngsdb mamba=0.1.2 python=3.8 -c conda-forge
conda activate ngsdb
mamba env update --file ngspipe/envs/requirements_ngsdb.yaml --prune
</code></pre>
<h3 id="2-convert-analysis-result-to-sqlite3-file">2. convert analysis result to sqlite3 file <a name="Table2Sqlite3"></a></h3>
<p>First, edit <code>ngspipe/config/ngsdb.config.yaml</code> file:</p>
<pre><code class="language-yaml"># path relative to where you run snakemake

# gene structure annotation
genomeAnno: &quot;testdata/GRCm38.83.chr19.gtf&quot; # gene annotation file, can be gtf or gff
genomeFasta: &quot;testdata/chr19.fa&quot; # genome sequence

# result directory of NGSPipe
resultsDir: &quot;results/result&quot;
reportsDir: &quot;results/report&quot;
dbDir: &quot;results/sqlite3&quot;

djangoCode: &quot;ngsdb&quot;
exp_data: &quot;results/result/quantify/quantify_by_stringtie/gene_fpkm_all_samples.tsv&quot;
report_path: &quot;&quot;
gbrowse_data: &quot;&quot;

# sample list file
samplesList: &quot;testdata/samples.xls&quot;
# sample file directory
samplesDir: &quot;testdata&quot;
# fastq suffix, read1
read1Suffix: &quot;_R1.fq.gz&quot;
read2Suffix: &quot;_R2.fq.gz&quot;
# replict can by 1,2,3
replict_num: 3
# condition for differential expression by deseq2
</code></pre>
<p>Second, edit <code>ngspipe/workflow/db_generate.Snakefile.py</code> file:</p>
<pre><code class="language-python"># detail parameters in pipe #

# 5. quantification
quantify_method = 'stringtie' # htseqcounts or featurecounts
quantify_outdir = join(config[&quot;resultsDir&quot;], &quot;quantify&quot;, &quot;quantify_by_{}&quot;.format(quantify_method))

# 4. transcript assembly
transcript_assembly_method = 'stringtie' # star
transcript_assembly_outdir = join(config[&quot;resultsDir&quot;], &quot;transcript_assembly&quot;, &quot;transcript_assembly_by_{}&quot;.format(transcript_assembly_method))


# blast
# 1. expression matrix database create
exp_db_outdir = join(config[&quot;dbDir&quot;], &quot;exp&quot;)
anno_db_outdir = join(config[&quot;dbDir&quot;], &quot;anno&quot;)

# sqlite3
# 2. blastdb
blastdb_outdir = join(config[&quot;dbDir&quot;], &quot;blastdb&quot;)

# 3. gffutils
gffdb_outdir = join(config[&quot;dbDir&quot;], &quot;gff_sqlite3&quot;)

# 4. genomebrowse
gbrowse_outdir = join(config[&quot;dbDir&quot;], &quot;gbrowse&quot;)
annotation_gbrowse_outdir = join(gbrowse_outdir, 'annotation')

# include modules
include: join(&quot;rules&quot;, &quot;8.db_generate_of_exp.Snakefile.py&quot;)
include: join(&quot;rules&quot;, &quot;8.db_generate_of_gff.Snakefile.py&quot;)
include: join(&quot;rules&quot;, &quot;8.db_generate_of_blastdb.Snakefile.py&quot;)
include: join(&quot;rules&quot;, &quot;8.db_generate_of_genomebrowser.Snakefile.py&quot;)
</code></pre>
<p>Then, run db_generate workflow to generate database files.</p>
<pre><code>snakemake -s ngspipe/workflow/db_generate.Snakefile.py --configfile ngspipe/config/ngsdb.config.yaml -p -j 1
</code></pre>
<h3 id="3-config">3. config <a name="DatabaseConfig"></a></h3>
<p>edit <code>mysite/mysite/settings.py</code> file</p>
<pre><code class="language-shell"># Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    # add custom app
    'geneAnno', # gene annotation from nr/nt/pfam/go/kegg
    'geneExpAtlas', # gene expression matrix
    'blast', # blast tool
]
</code></pre>
<h3 id="4-start-server">4. start server <a name="RunServer"></a></h3>
<p>Starting server by run <code>python ngsdb/manage.py runserver</code>, visit sebsite on http://127.0.0.1:8000.</p>
<h3 id="5-add-your-custome-data">5. add your custome data</h3>
<ol>
<li>table data in a page</li>
<li>
<p>custome script in wooey</p>
<p>python manage.py makemigrations
python manage.py migrate
python manage.py addscript ../ngsdb/tools/test1.py</p>
</li>
</ol>
<p>If you want use wooey tools on a task model <code>celery -A ProjectName worker -c 1 --beat -l info</code> optional.</p>
<h2 id="overview-the-ngsdb-apps">overview the NGSDb apps</h2>
<p>We use django project to constructed our NGSDb. We have pareparied many apps for you. embeed. Please have a look:</p>
<table>
<thead>
<tr>
<th>django apps</th>
<th>description</th>
<th>package</th>
</tr>
</thead>
<tbody>
<tr>
<td>report</td>
<td>ngs analysis report</td>
<td>snakemake</td>
</tr>
<tr>
<td>geneExpAtlas</td>
<td>table</td>
<td>datatable js</td>
</tr>
<tr>
<td>network</td>
<td>table</td>
<td>datatable js</td>
</tr>
<tr>
<td>igv</td>
<td>genome browse</td>
<td><a href="https://github.com/igvteam/igv.js/wiki">IGV</a></td>
</tr>
<tr>
<td>blastplus</td>
<td>ncbi blast +</td>
<td>NCBI</td>
</tr>
<tr>
<td>ngstools</td>
<td>python script</td>
<td>wooey</td>
</tr>
<tr>
<td>efp</td>
<td>efp browse</td>
<td></td>
</tr>
<tr>
<td>heatmap</td>
<td>expression</td>
<td>clustergrammer-js https://clustergrammer.readthedocs.io/clustergrammer_js.html</td>
</tr>
</tbody>
</table>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../linux/" class="btn btn-neutral float-right" title="Linux & shell basic">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../NGSPipe-resequecing/" class="btn btn-neutral" title="NGSPipe resequecing"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
      <p>Bioinformatics Group, XTBG, 2021</p>
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/xuanblo/NGSPipeDb/mkdocs/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../NGSPipe-resequecing/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../linux/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
